src/dotnet/commands/dotnet-crossgen/CrossGenContext.cs

```diff
@@ -0,0 +1,243 @@
+            }
+            else
+            {
+                ridList = ridList.Concat(ridFallback.Fallbacks);
+            }
+
+            var arch = RuntimeEnvironment.RuntimeArchitecture;
+            var diaSymReaderFileName = $"{DynamicLibPrefix}Microsoft.DiaSymReader.Native.{arch}{DynamicLibSuffix}";
+            var probeLocations = new string[] { Path.Combine(ApplicationEnvironment.ApplicationBasePath, diaSymReaderFileName) }.Concat(
+                    ridList.Select(rid => Path.Combine(ApplicationEnvironment.ApplicationBasePath, "runtimes", rid, "native", diaSymReaderFileName)));
+
+            // x64, aka amd64
+            if (arch == "x64")
+            {
+                var archSuffix = $".{arch}{DynamicLibSuffix}";
+                probeLocations = probeLocations.Concat(
+                                    probeLocations.Where(l => l.EndsWith(archSuffix))
+                                                    .Select(l => l.Substring(0, l.Length - archSuffix.Length) + $".amd64{DynamicLibSuffix}"));
+            }
+
+            var foundLocation = probeLocations.FirstOrDefault(l => File.Exists(l));
+
+            if (foundLocation == null)
+            {
+                throw new CrossGenException($"Failed to locate DiaSymReader for runtime {targetRid}");
+            }
+
+            Reporter.Verbose.WriteLine($"Found DiaSymReader {foundLocation}");
+            return foundLocation;
+        }
+
+        private string LocateSharedFramework(RuntimeConfigFramework framework)
+        {
+            var dotnetHome = Path.GetDirectoryName(new Muxer().MuxerPath);
+            var shareFrameworksDir = Path.Combine(dotnetHome, "shared", framework.Name);
+
+            if (!Directory.Exists(shareFrameworksDir))
+            {
+                throw new CrossGenException($"Shared framework {shareFrameworksDir} does not exist");
+            }
+
+            var version = framework.Version;
+            var exactMatch = Path.Combine(shareFrameworksDir, version);
+            if (Directory.Exists(exactMatch))
+            {
+                return exactMatch;
+            }
+            else
+            {
+                Reporter.Verbose.WriteLine($"Cannot find shared framework in: {exactMatch}, trying to auto roll forward.");
+                return AutoRollForward(shareFrameworksDir, version);
+            }
+        }
+
+        private string AutoRollForward(string root, string version)
+        {
+            var targetVersion = NuGetVersion.Parse(version);
+            var candidateNames = Directory.GetDirectories(root).Select(d => Path.GetFileName(d));
+
+            string bestMatch = null;
+            NuGetVersion bestMatchVersion = null;
+            foreach (var candidateName in candidateNames)
+            {
+                var currentVersion = NuGetVersion.Parse(candidateName);
+                if (currentVersion.Major == targetVersion.Major && currentVersion.CompareTo(targetVersion) > 0)
```

Comment:
The dotnet host only rolls forward on patch versions (Major.minor.patch).  Do we want to keep the same behavior in crossgen?  Or are we saying that any major versions of the shared framework will be compatible with crossgen?

/cc @gkhanna79